src/js/components/SchemaField.js

```diff
@@ -0,0 +1,306 @@
+        className={"field-select-narrow"}
+        name={name}
+        type="text"
+        value={formData}
+        onChange={event => onChange(event.target.value)}
+        onBlur={onBlur && (event => onBlur(name, event.target.value))}
+        onFocus={onFocus && (event => onFocus(name, event.target.value))}
+      >
+        {options}
+      </FieldSelect>
+    );
+    if (autofocus) {
+      field = (
+        <FieldAutofocus>
+          {field}
+        </FieldAutofocus>
+      );
+    }
+
+    return (
+      <div>
+        <FieldLabel>
+          {this.getFieldHeading(required, name, schema.description)}
+        </FieldLabel>
+        {field}
+        <FieldError>{errorMessage}</FieldError>
+      </div>
+    );
+  }
+
+  renderTextInput(errorMessage, autofocus, props) {
+    const {
+      required,
+      name,
+      schema,
+      formData,
+      onChange,
+      onBlur,
+      onFocus
+    } = props;
+
+    let field = (
+      <FieldInput
+        id={name}
+        type={"text"}
+        className={"field-input-text-narrow"}
+        autoFocus={autofocus}
+        name={name}
+        value={formData}
+        onChange={event => onChange(event.target.value)}
+        onBlur={onBlur && (event => onBlur(name, event.target.value))}
+        onFocus={onFocus && (event => onFocus(name, event.target.value))}
+      />
+    );
+    if (autofocus) {
+      field = (
+        <FieldAutofocus>
+          {field}
+        </FieldAutofocus>
+      );
+    }
+
+    return (
+      <div>
+        <FieldLabel>
+          {this.getFieldHeading(required, name, schema.description)}
+        </FieldLabel>
+        {field}
+        <FieldError>{errorMessage}</FieldError>
+      </div>
+    );
+  }
+
+  renderNumberInput(errorMessage, autofocus, props) {
+    const {
+      required,
+      name,
+      schema,
+      formData,
+      onChange,
+      onBlur,
+      onFocus
+    } = props;
+
+    const handleChange = event => {
+      const value = event.target.value;
+
+      if (value === "") {
+        onChange("");
+      } else {
+        onChange(Number(value));
+      }
+    };
+
+    let field = (
+      <FieldInput
+        id={name}
+        type={"number"}
+        className={"field-input-number-narrow"}
+        autoFocus={autofocus}
+        name={name}
+        value={formData}
+        onChange={handleChange.bind(this)}
+        onBlur={onBlur && (event => onBlur(name, event.target.value))}
+        onFocus={onFocus && (event => onFocus(name, event.target.value))}
+      />
+    );
+    if (autofocus) {
+      field = (
+        <FieldAutofocus>
+          {field}
+        </FieldAutofocus>
+      );
+    }
+
+    return (
+      <div>
+        <FieldLabel>
+          {this.getFieldHeading(required, name, schema.description)}
+        </FieldLabel>
+        {field}
+        <FieldError>{errorMessage}</FieldError>
+      </div>
+    );
+  }
+
+  getFieldHeading(required, name = "", description) {
+    let asterisk = null;
```

Comment:
Very personal opinion about the name.
I would rather use something like `requiredSign` because if we decide to change the sign from an asterisk to an exclamation mark we also need to rename the variable. So I would rather opt for a more generic name which is describing what it does and not how it looks like.

Feel free to adjust.

Reply:
I agree with that reasoning 👍